Support mapping time series collections (#2687)
doctrine/mongodb-odm: docs/en/cookbook/time-series-data.rst

    and rarely changes

-
    Measurements, which are the data points tracked at increments in time.

.. note::

    Support for time series collections was added in MongoDB 5.0. Attempting to
    use this functionality on older server versions will result in an error on
    schema creation.

Creating The Model
------------------

For this example, we'll be storing data from multiple temperature sensors. Other
examples for time series include stock data, price information, website visitors,
and vehicle telemetry (speed, position, etc.).

First, we define the model for our data:

.. code-block:: php

    <?php

    use DateTimeImmutable;
    use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
    use MongoDB\BSON\ObjectId;

    #[ODM\Document]
    readonly class Measurement
    {
        #[ODM\Id]
        public string $id;

        public function __construct(
            #[ODM\Field(type: 'date_immutable')]
            public DateTimeImmutable $time,
            #[ODM\Field(type: 'int')]
            public int $sensorId,
            #[ODM\Field(type: 'float')]
            public float $temperature,
        ) {

>             $this->id = (string) new ObjectId();

Reviewer: Does ODM require a string cast here, or is it only done to satisfy the `string` type hint on the property? IIRC, ODM would have allowed an ObjectId on the initially created entity, and left is as-is for insertion.

Author: The default `id` type converts an `ObjectId` instance to a string in the entity. It's bothered me for a while as working with the `ObjectId` instance directly would make more sense, but I've shied away from making the big BC break. We could use `ObjectId|string` as type for the identifier, but this would create ambiguity in the API as users would always have to check types when using the identifier, hence the string cast.

        }
    }

Note that we defined the entire model as readonly. While we could theoretically
change values in the document, in this example we'll assume that the data will
not change.

Now we can mark the document as a time series document. This is done using the
``TimeSeries`` attribute. Since we'll be storing data from multiple sensors, we
store the ID of each sensor as metadata. We only use the temperature as a
measurement, but we could also add additional sensors. With that in mind, we can
add the ``TimeSeries`` attribute:

.. code-block:: php

    <?php

    // ...

    #[ODM\Document]
    #[ODM\TimeSeries(timeField: 'time', metaField: 'sensorId')]
    readonly class Measurement
    {
        // ...
    }

> Once we created the schema, we can store our measurements in this time series

Reviewer: ```suggestion
Once we create the schema, we can store our measurements in this time series
```

Alternatively: "once we have created the schema"

> collection and let MongoDB optimise the storage for faster queries:

Reviewer: ```suggestion
collection and let MongoDB optimize the storage for faster queries:
```

Decide if ODM docs use American or British English for spelling.

.. code-block:: php

    <?php

    $measurement = new Measurement(
        time: new DateTimeImmutable(),
        sensorId: $sensorId,
        temperature: $temperature,
    );

    $documentManager->persist($measurement);
    $documentManager->flush();

> Note that other functionality, such as querying, aggregating data using
> aggregation pipeline, or removing data works the same as with other collections.

Reviewer: ```suggestion
Note that other functionality such as querying, using aggregation pipelines, or
removing data works the same as with other collections.
```

Removing word repetition here.

If you preserve the comma after "functionality", then there should be a second comma in "removing data, works". Both commas would segregate the "such as" list.

Considerations
--------------

With the mapping above, data is stored with a granularity of seconds. Depending
on how often measurements come in, we can reduce the granularity to minutes or
hours. This changes how the data is stored internally by changing the bucket
size. This affects storage requirements and query performance.

For example, with the default ``seconds`` granularity, each bucket groups
documents for one hour. If each sensor only reports data every few minutes, we'd

> do well to only store them with a ``minute`` granularity. This reduces the

Reviewer: ```suggestion
do well to configure ``minute`` granularity. This reduces the
```

Alternatively, replace "configure" with "opt for". Either way, we can remove "a" here.
